enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,634 @@
+         delete:
+```
+7. Users want to delete HTTP response header which need to be sent only to the client hitting
+   the route with the HTTP Response Header values.
+8. Users edits/creates the Route having Spec as follows -
+```yaml
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  labels:
+    type: sharded 
+  name: hello-openshift-edge
+  namespace: hello-openshift
+spec:
+  subdomain: hello-openshift 
+  tls:
+    termination: edge
+  to:
+    kind: Service
+    name: hello-openshift
+    httpHeaders:
+  headersManipulation:
+    response:
+    - name: X-Frame-Options
+      action: Delete
+      delete:
+    - name: X-XSS-Protection
+      action: Delete
+      delete:
+```
+#### Variation [optional]
+
+### API Extensions
+
+A) Set/Delete HTTP headers in the global section of haproxy.config via IngressController's Spec.
+The existing Ingress Controller API `HTTPHeaders *IngressControllerHTTPHeaders` is extended by adding an field for action to be performed,
+and field for headers to be set/deleted.
+
+```go
+// IngressControllerSpec is the specification of the desired behavior of the
+// IngressController.
+type IngressControllerSpec struct {
+    // httpHeaders defines policy for HTTP headers.
+    //
+    // If this field is empty, the default values are used.
+    //
+    // +optional
+    HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`
+}
+```
+
+```go
+// IngressControllerHTTPHeaders specifies how the IngressController handles
+// certain HTTP headers.
+type IngressControllerHTTPHeaders struct { 
+	// headersManipulation specifies options for set/delete headers.
+    HeadersManipulation IngressControllerHTTPHeadersManipulation `json:"headersManipulation,omitempty"`
+}
+```
+
+```go
+// IngressControllerHeadersManipulation defines whether we want to set Response or Request Headers.
+// Currently set/delete of HTTP Response Headers is available only.
+type IngressControllerHTTPHeadersManipulation struct {
+	Response []IngressControllerSetOrDeleteHTTPHeader `json:"response,omitempty"`
+}
+```
+
+```go
+// IngressControllerSetOrDeleteHTTPHeader specifies options for set/delete a header.
+type IngressControllerSetOrDeleteHTTPHeader struct {
+    // name specifies a header name to be set(replace)/append/delete.  Its value must be a valid HTTP header.
+    // name as defined in RFC 2616 section 4.2.
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=256
+    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
+    // +required
+    Name string `json:"name"`
+
+    IngressControllerSetOrDeleteHTTPHeaderUnion `json:",inline"`
+}
+```
+
+```go
+// IngressControllerSetOrDeleteHTTPHeaderUnion specifies fields to mention the header details and the action you want to perform on them.
+// +union
+type IngressControllerSetOrDeleteHTTPHeaderUnion struct {
+	// action lets you specify headers for Set or Delete.
```

Comment:
Actually the godoc should use the lowercase "action" to match the json tag (see https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#use-json-field-names-in-godoc).  Otherwise I like the suggestion.  
```suggestion
	// action defines the type of the action to be applied on the header.
```

Reply:
Interesting, didn't know this.
`go doc` (not `godoc`) seems to be fine with the case mismatch btw:
```
$ cat here/f.go 
package here

import (
	"fmt"
)

// here is a function
func Here() {
	fmt.Println("here")
}

$ go doc -all here
package here // import "github.com/alebedev87/dummy/here"


FUNCTIONS

func Here()
    here is a function
```
Not sure which one is used in pkg.go.dev, for instance.